### PR TITLE
feat: adding api calls for cart and wishlist

### PR DESCRIPTION
**Pull request by mariyasada**
-adding api calls for cart and wishlist management
-user can add products to the cart and wishlist
-user can add products to the cart from the wishlist
-user can remove the product from the wishlist
-user can increase and decrease the quantity of products in the cart
- calculating bill  of products
- address management(add,delete,edit address) of user
- logout functionality

https://user-images.githubusercontent.com/77275100/163719670-c20b6d87-afbf-484d-b30b-88c7909e84b5.mp4

### CODE REVIEW

**Review by sameep-s — APPROVED** [review]
Things that i liked:
- good folder structure
- good naming

File: src/components/categories/Categories.jsx [hunk 1]
```diff
@@ -1,37 +1,41 @@
...
-            key={category._id}>
-                {category.categoryName}
+      {categories.map((category) => {
+        return (
+          <div className="category-container" key={category.id}>
+            <Link
+              to={`/productpage?category=${category.categoryName}`}
+              onClick={
+                category.categoryName === "Men"
+                  ? () => dispatch({ type: "SORT_BY_MEN_CATEGORY" })
```
Comment (paarthbajaj) [inline src/components/categories/Categories.jsx:29]: Good use of chaining. Good effort.

File: src/context/CartAndWishlist-context.js [hunk 2]
```diff
@@ -1,28 +1,119 @@
...
+
+  // ADD ITEM TO WISHLIST
+  const addProductToWishlist =async(product,setBtnDisabled)=>{
+      const {data,status}=await AddItemToWishList(product,user,setBtnDisabled);          
+      if(status===201)
+      {
+          wishListDispatch({type:"ADD_ITEM_TO_WISHLIST",payload:data.wishlist})
+      }
+      else{
+          throw new Error("could not complete the request");
```
Comment (paarthbajaj) [inline src/context/CartAndWishlist-context.js:40]: Very good error handling, I liked it.

File: src/Router/RequiresAuth.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import React from 'react'
+import { Navigate, useLocation } from 'react-router-dom';
+import { useAuth } from '../context/Authentication/auth-context';
+
+
+export const RequiresAuth = ({children}) => {
+    const {user:{isloggedIn}}=useAuth();   
+    const location=useLocation(); 
+
+  return isloggedIn? children:<Navigate  to="/login" replace state={{from:location}}/>
```
Comment (paarthbajaj) [inline src/Router/RequiresAuth.js:10]: Good use of advance react routers as well.

**Review by paarthbajaj — APPROVED** [review]
Could not find any need of improvement. You have coded with all the logics well built and good use of advanced react. Well done. You can merge this PR.

File: src/context/utils/cartUtils.js [hunk 4]
```diff
@@ -1,15 +1,12 @@
...
+let isExist =false;
+ state.cartItem.map((item)=>{
+      if(item._id === product._id) {
+          return isExist=true
+       }
+       else{
+           return isExist;
+       }
+    })
+        return isExist;
```
Comment (JantuDeb) [inline src/context/utils/cartUtils.js:11]: `state.cartItem.some(item=>item._id===product._id)`

**Review by JantuDeb — APPROVED** [review]
overall the code looks fine

**Pull request merged**